modyn/storage/src/internal/grpc/storage_service_impl.cpp

```diff
@@ -0,0 +1,596 @@
+    T response;
+    for (auto& row : rs) {
+      response.add_keys(row.get<long long>(0));    // NOLINT google-runtime-int
+      response.add_labels(row.get<long long>(1));  // NOLINT google-runtime-int
+    }
+    writer->Write(response);
+  } else {
+    // If the number of samples is greater than the sample batch size, retrieve the samples in batches of size
+    // sample_batch_size_. The batches are retrieved asynchronously and the futures are stored in a queue. When the
+    // queue is full, the first future is waited for and the response is sent to the client. This is repeated until all
+    // of the futures have been waited for.
+    std::queue<std::future<SampleData>> sample_ids_futures_queue;
+
+    for (int64_t i = 0; i < number_of_samples; i += sample_batch_size_) {
+      if (static_cast<int64_t>(sample_ids_futures_queue.size()) == retrieval_threads_) {
+        // The queue is full, wait for the first future to finish and send the response.
+        T response;
+
+        SampleData sample_data = sample_ids_futures_queue.front().get();
+        sample_ids_futures_queue.pop();
+
+        for (size_t i = 0; i < sample_data.ids.size(); i++) {
+          response.add_keys(sample_data.ids[i]);
+          response.add_labels(sample_data.labels[i]);
+        }
+
+        writer->Write(response);
+      }
+
+      // Start a new future to retrieve the next batch of samples.
+      std::future<SampleData> sample_ids_future =
+          std::async(std::launch::async, get_sample_subset, file_id, i, i + sample_batch_size_ - 1,  // NOLINT
+                     std::ref(storage_database_connection_));
+      sample_ids_futures_queue.push(std::move(sample_ids_future));
+    }
+
+    // Wait for all of the futures to finish executing before returning.
+    while (!sample_ids_futures_queue.empty()) {
+      T response;
+
+      SampleData sample_data = sample_ids_futures_queue.front().get();
+      sample_ids_futures_queue.pop();
+
+      for (size_t i = 0; i < sample_data.ids.size(); i++) {
+        response.add_keys(sample_data.ids[i]);
+        response.add_labels(sample_data.labels[i]);
+      }
+
+      writer->Write(response);
+    }
+  }
+}
+
+SampleData StorageServiceImpl::get_sample_subset(
+    int64_t file_id, int64_t start_index, int64_t end_index,
+    const storage::database::StorageDatabaseConnection& storage_database_connection) {
+  soci::session session = storage_database_connection.get_session();
+  const int64_t number_of_samples = end_index - start_index + 1;
+  std::vector<int64_t> sample_ids(number_of_samples + 1);
+  std::vector<int64_t> sample_labels(number_of_samples + 1);
+  session << "SELECT sample_id, label FROM samples WHERE file_id = :file_id AND sample_index >= :start_index AND "
+             "sample_index "
+             "<= :end_index",
+      soci::into(sample_ids), soci::into(sample_labels), soci::use(file_id), soci::use(start_index),
```

Comment:
here, we are not using the rowset interface. we should make this consistent. The parallel retrieval with multiple threads (although I do not understand the future logic yet TBH) is orthogonal to partitioned execution.

Reply:
How is this orthogonal? We are retrieving a subset (partition) of the samples, the retrieval being the heavy operation, and return each subset as a yield to the writer.